controllers/product.go

```diff
@@ -0,0 +1,174 @@
+
+	"github.com/cloudinary/cloudinary-go/api/uploader"
+	"github.com/gin-gonic/gin"
+)
+
+//ProductController ...
+type ProductController struct{}
+
+func UploadProductImage(img_path string) (url string, err error) {
+
+	if _, err := os.Stat(img_path); os.IsNotExist(err) {
+		return "", err
+	}
+
+	var ctx = context.Background()
+	uploadResult, err := util.CLD.Upload.Upload(ctx, img_path, uploader.UploadParams{Folder: os.ExpandEnv("CLOUDAINARY_STORAGE_FOLDER")})
+	if err != nil {
+		return "", err
+	}
+
+	return uploadResult.SecureURL, nil
+}
+
+func (ctrl ProductController) AddProduct(c *gin.Context) {
+
+	// Get the email of the user to compare with email in the header
+	var input models.AddProductInput
+	err := c.ShouldBind(&input)
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": err.Error(), "success": false})
+		return
+	}
+
+	var user models.User
+	if db.DB.Find(&user, "id=?", input.UserID).RecordNotFound() {
+		c.AbortWithStatusJSON(http.StatusNotFound, gin.H{"error": "Something Went Wrong", "success": false})
+		return
+	}
+
+	email := c.Request.Header.Get("email")
+	if email != user.Email {
+		c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "You are not authorized to perform this action", "success": false})
+		c.Abort()
+		return
+	}
+
+	// Extract image from the form
+	r := c.Request
+	r.ParseMultipartForm(10 << 20)
+
+	file, _, err := r.FormFile("image")
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": err.Error(), "success": false})
+		return
+	}
+	defer file.Close()
+
+	// Store image in the server side
+	tempFile, err := ioutil.TempFile("res", "upload-*.png")
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, gin.H{"error": err.Error(), "success": false})
+	}
+	defer tempFile.Close()
+
+	fileBytes, err := ioutil.ReadAll(file)
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, gin.H{"error": err.Error(), "success": false})
+	}
+	// write this byte array to our temporary file
+	tempFile.Write(fileBytes)
+
+	// upload the image to cloudinary cloud
+	img_url, err := UploadProductImage(tempFile.Name())
+	if err != nil {
+		c.IndentedJSON(http.StatusInternalServerError, gin.H{"error": err.Error(), "success": false})
+		return
+	}
+
+	// remove the image from the server side after uploading it
+	e := os.Remove(tempFile.Name())
+	if e != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, gin.H{"error": e.Error(), "success": false})
+	}
+
+	// Store the new product in the database
+	product := models.Product{UserID: input.UserID, Title: input.Title, Content: input.Content, Price: input.Price, ImageURL: img_url}
+	db.DB.Create(&product)
+
+	c.IndentedJSON(http.StatusOK, gin.H{"data": product, "success": true})
+}
+
+func (ctrl ProductController) GetAll(c *gin.Context) {
+	var products []models.Product
+	db.DB.Find(&products)
+
+	c.IndentedJSON(http.StatusOK, gin.H{"data": products, "sucess": true})
```

Comment:
Typo.
**success** not  _sucess_ .
I mean, Make it consistent so the frontend can determine what to check.
Probably its useless since we already send HttpStatus code?